prettier/src/utils/tokenize.ts

```diff
@@ -0,0 +1,160 @@
+const tokenPlaceholderMap = new Map();
+let tokenIndex = 0;
+
+const getTokenByValue = (match, type) => {
+  let tokens = tokenPlaceholderMap.entries();
+  for (let token of tokens) {
+    if (token[1].type === type && token[1].token === match) {
+      return token[0];
+    }
+  }
+};
+
+export const tokenize = (input) => {
+  const HUBL_COMMENT_REGEX = /{#.*?#}/gms;
+  const HUBL_TAG_REGEX = /({%.+?%})/gs;
+  const LINE_BREAK_REGEX = /[\r\n]+/gm;
+  const HUBL_VARIABLE_REGEX = /({{.+?}})/gs;
+  const HTML_TAG_WITH_HUBL_TAG_REGEX = /<[^>]*?(?={%|{{).*?>/gms;
+  const STYLE_BLOCK_WITH_HUBL_REGEX = /<style.[^>]*?(?={%|{{).*?style>/gms;
+  const JSON_BLOCK_REGEX =
+    /(?<={% widget_attribute.*is_json="?true"? %}|{% module_attribute.*is_json="?true"? %}).*?(?={%.*?end_module_attribute.*?%}|{%.*?end_widget_attribute.*?%})/gims;
+
+  const TOKEN_TYPES = {
+    nestedHTML: "nestedHTML",
+    nestedStyleTag: "nestedStyleTag",
+    jsonBlock: "jsonBlock",
+    hublComment: "hublComment",
+    hublTag: "hublTag",
+    hublVariable: "hublVariable",
+  };
+
+  const tokenTypes = {
+    nestedHTML: {
+      placeholder: "npe$t_",
```

Comment:
We're just doing a replace on `$t` in these placeholders. Alternatively, this could be a function like `tokenReplacer`. This felt a little cleaner, but open to feedback!